pkg/ottl/ottlfuncs/func_to_key_value_string.go

```diff
@@ -0,0 +1,122 @@
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ToKeyValueStringArguments[K any] struct {
+	Target        ottl.PMapGetter[K]
+	Delimiter     ottl.Optional[string]
+	PairDelimiter ottl.Optional[string]
+	SortOutput    ottl.Optional[bool]
+}
+
+func NewToKeyValueStringFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ToKeyValueString", &ToKeyValueStringArguments[K]{}, createToKeyValueStringFunction[K])
+}
+
+func createToKeyValueStringFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ToKeyValueStringArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ToKeyValueStringFactory args must be of type *ToKeyValueStringArguments[K]")
+	}
+
+	return toKeyValueString[K](args.Target, args.Delimiter, args.PairDelimiter, args.SortOutput)
+}
+
+func toKeyValueString[K any](target ottl.PMapGetter[K], d ottl.Optional[string], p ottl.Optional[string], s ottl.Optional[bool]) (ottl.ExprFunc[K], error) {
+	delimiter := "="
+	if !d.IsEmpty() {
+		if d.Get() == "" {
+			return nil, fmt.Errorf("delimiter cannot be set to an empty string")
+		}
+		delimiter = d.Get()
+	}
+
+	pairDelimiter := " "
+	if !p.IsEmpty() {
+		if p.Get() == "" {
+			return nil, fmt.Errorf("pair delimiter cannot be set to an empty string")
+		}
+		pairDelimiter = p.Get()
+	}
+
+	if pairDelimiter == delimiter {
+		return nil, fmt.Errorf("pair delimiter %q cannot be equal to delimiter %q", pairDelimiter, delimiter)
+	}
+
+	sortOutput := false
+	if !s.IsEmpty() {
+		sortOutput = s.Get()
+	}
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		source, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		kvString := convertMapToKV(source, delimiter, pairDelimiter, sortOutput)
+
+		return kvString, nil
+	}, nil
+}
+
+// convertMapToKV converts a pcommon.Map to a key value string
+func convertMapToKV(target pcommon.Map, delimiter string, pairDelimiter string, sortOutput bool) string {
+
+	var kvStrings []string
+	if sortOutput {
+		var keyValues []struct {
+			key string
+			val pcommon.Value
+		}
+
+		// Sort by keys
+		target.Range(func(k string, v pcommon.Value) bool {
+			keyValues = append(keyValues, struct {
+				key string
+				val pcommon.Value
+			}{key: k, val: v})
+			return true
+		})
+		gosort.Slice(keyValues, func(i, j int) bool {
+			return keyValues[i].key < keyValues[j].key
+		})
+
+		// Convert KV pairs
+		for _, kv := range keyValues {
+			k := escapeAndQuoteKV(kv.key, delimiter, pairDelimiter)
+			vStr := escapeAndQuoteKV(kv.val.AsString(), delimiter, pairDelimiter)
+			kvStrings = append(kvStrings, fmt.Sprintf("%s%s%v", k, delimiter, vStr))
+		}
+	} else {
+		target.Range(func(k string, v pcommon.Value) bool {
+			key := escapeAndQuoteKV(k, delimiter, pairDelimiter)
+			vStr := escapeAndQuoteKV(v.AsString(), delimiter, pairDelimiter)
+			kvStrings = append(kvStrings, fmt.Sprintf("%s%s%v", key, delimiter, vStr))
+			return true
```

Comment:
It would be nice to reuse this logic so that in the future, if we need to change it, we don't accidentally change it for only the sorted path or only the unsorted path.